### PR TITLE
Fix credentials serialization with publishing

### PR DESCRIPTION
**Pull request by rieske**
Fixes https://github.com/gradle/gradle/issues/14902

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/credentials/AwsCredentials.java [hunk 1]
```diff
@@ -37,6 +41,7 @@
     /**
      * Returns the secret key to use to authenticate with AWS.
      */
+    @Input
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/credentials/AwsCredentials.java:44]: The secret key should _not_ be an input to the task.

File: subprojects/core-api/src/main/java/org/gradle/api/credentials/PasswordCredentials.java [hunk 2]
```diff
@@ -44,6 +47,7 @@
      *
      * @return The password. May be null.
      */
+    @Input
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/credentials/PasswordCredentials.java:50]: The password should _not_ be an input to the task.

**Review by wolfs — CHANGES_REQUESTED** [review]
Please make sure that secret values aren't inputs. The md5 hashes of those will be stored in the build cache and so could be easily obtained.

File: subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultPasswordCredentials.java [hunk 3]
```diff
@@ -29,6 +30,7 @@ public DefaultPasswordCredentials(String username, String password) {
         this.password = password;
     }
 
+    @Internal
```
Comment (wolfs) [inline subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultPasswordCredentials.java:33]: You could add the username as an input. It shouldn't be confidential.

File: subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultHttpHeaderCredentials.java [hunk 4]
```diff
@@ -16,12 +16,14 @@
...
 
 import org.gradle.api.credentials.HttpHeaderCredentials;
+import org.gradle.api.tasks.Internal;
 
 public class DefaultHttpHeaderCredentials implements HttpHeaderCredentials {
 
     private String name;
     private String value;
 
+    @Internal
```
Comment (wolfs) [inline subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultHttpHeaderCredentials.java:26]: Is the header name confidential? It probably isn't, so I would mark it as `@Input`. Note that the header name can be null, so you also need `@Optional`.

**Review by wolfs — APPROVED** [review]
LGTM! You could add more properties as inputs, though that isn't necessary.

File: subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultPasswordCredentials.java [hunk 5]
```diff
@@ -39,6 +42,7 @@ public void setUsername(String username) {
         this.username = username;
     }
 
+    @Internal
```
Comment (JLLeitschuh) [inline subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultPasswordCredentials.java:45]: Won't this mean that if the password changes, it won't force the task to be marked as out-of-date?

Comment (wolfs) [reply subprojects/model-core/src/main/java/org/gradle/internal/credentials/DefaultPasswordCredentials.java:45]: It does, and that is fine. Passwords do not change the outputs of the task, only if it succeeds or not. So if you already have a successful execution with one password, why would you need to run the same thing again with a different password?

**Comment by JLLeitschuh** [comment]
@wolfs couldn't we use a SHA-512 hash of the secret as an input? That way the secret can be hashed but kept secret?

**Comment by wolfs** [comment]
> @wolfs couldn't we use a SHA-512 hash of the secret as an input? That way the secret can be hashed but kept secret?

We could. I still believe secrets shouldn't be an input to the task, because they don't affect the outputs.

**Comment by JLLeitschuh** [comment]
@wolfs but they do impact the up-to-date check. Is there a good way that we could still convey the change in a credential needs to force a task to be marked as out-of-date?

**Comment by wolfs** [comment]
In my opinion, credentials are like memory settings and fork options. If you change the memory settings for a test task, it may change from failing to passing, though if it passes, then the actual outcome of the task is the same independent of the memory settings. So changing the memory settings only do not influence the outputs of a task.

Same goes for credentials: If those are an input, that would mean that the task would produce different output if it passes depending on the credentials. Though it shouldn’t: If it passes, the credentials have been correct, and the outputs of the task should be the same no matter which validate credentials you use.

**Pull request merged**